torch/nn/modules/_functions.py

```diff
@@ -16,17 +17,25 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
         if size == 1 and world_size < 2:
             raise ValueError('Expected more than 1 value per channel when training, got input size {}'.format(size))
 
-        # calculate mean/invstd for input.
-        mean, invstd = torch.batch_norm_stats(input, eps)
+        num_channels = input.shape[1]
+        if input.numel() > 0:
+            # calculate mean/invstd for input.
+            mean, invstd = torch.batch_norm_stats(input, eps)
 
-        count = torch.full((1,), input.numel() // input.size(1),
-                           dtype=mean.dtype,
-                           device=mean.device)
+            count = torch.full((1,), input.numel() // input.size(1),
+                            dtype=mean.dtype,
+                            device=mean.device)
 
+            # C, C, 1 -> (2C + 1)
+            combined = torch.cat([mean, invstd, count], dim=0)
+        else:
+            # for empty input, directly set all stats to 0
+            combined = torch.zeros(
```

Comment:
Wouldn't something like: `torch.zeros(dtype=input.dtype, device=input.device).expand(2 * num_channels + 1)` also work and reduce the bandwidth that is wasted?
Not sure how the rpc is handling non-contiguous Tensors.

Reply:
```
torch.zeros(dtype=input.dtype, device=input.device).expand(2 * num_channels + 1)
```

Curious, what bandwidth does the above code save? And why RPC is relevant here?

Reply:
Oh I see. Not sure if this gonna work. Collectives use ProcessGroup and will call NCCL APIs under the hood. IIRC, NCCL expects contiguous tensors and will directly read `numel()` elements from the memory pointer. Let me double check on that

Reply:
```
  File "/raid/shenli/pytorch/torch/distributed/distributed_c10d.py", line 2130, in _all_gather_base                                                             
    work = group._allgather_base(output_tensor, input_tensor)                                                                                                   
RuntimeError: Tensors must be contiguous                                                                                                                        
Exception raised from check_gpu_single_tensor at ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227 (most recent call first):  
```

Hit the above error, caused by the following line.  

https://github.com/pytorch/pytorch/blob/835cc66e5dd26db558931b4fe47b45e08a3a09f7/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1226-L1228

Reply:
Ok then.
As a side note, I think you should look into that as it is potentially a major bandwidth gain (and if I understand correctly, this is an expensive commodity).